tests/networkpolicy_test.go

```diff
@@ -1,173 +1,234 @@
-	var virtClient kubecli.KubevirtClient
+	var (
+		virtClient                                          kubecli.KubevirtClient
+		serverVMI, clientVMI, clientVMIAlternativeNamespace *v1.VirtualMachineInstance
+		serverVMILabels                                     = map[string]string{"type": "test"}
+	)
+	Context("when three cirros VMs with default networking are started and serverVMI start an HTTP server on port 80 and 81", func() {
+		tests.BeforeAll(func() {
+			var err error
 
-	var vmia *v1.VirtualMachineInstance
-	var vmib *v1.VirtualMachineInstance
-	var vmic *v1.VirtualMachineInstance
+			virtClient, err = kubecli.GetKubevirtClient()
+			tests.PanicOnError(err)
 
-	tests.BeforeAll(func() {
-		var err error
+			tests.SkipIfUseFlannel(virtClient)
+			skipNetworkPolicyRunningOnKindInfra()
+			tests.BeforeTestCleanup()
 
-		virtClient, err = kubecli.GetKubevirtClient()
-		tests.PanicOnError(err)
+			// Create three vmis, serverVMI and clientVMI are in same namespace, clientVMIAlternativeNamespace is in different namespace
+			serverVMI = createVMICirros(virtClient, tests.NamespaceTestDefault, serverVMILabels, libvmi.WithInterface(
+				libvmi.InterfaceDeviceWithMasqueradeBinding(
+					v1.Port{
+						Name:     "http80",
+						Port:     80,
+						Protocol: "TCP",
+					},
+					v1.Port{
+						Name:     "http81",
+						Port:     81,
+						Protocol: "TCP",
+					},
+				),
+			))
+			clientVMI = createVMICirros(virtClient, tests.NamespaceTestDefault, map[string]string{}, libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()))
+			clientVMIAlternativeNamespace = createVMICirros(virtClient, tests.NamespaceTestAlternative, map[string]string{}, libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()))
 
-		tests.SkipIfUseFlannel(virtClient)
-		skipNetworkPolicyRunningOnKindInfra()
-		tests.BeforeTestCleanup()
+			serverVMI = tests.WaitUntilVMIReady(serverVMI, tests.LoggedInCirrosExpecter)
+			clientVMI = tests.WaitUntilVMIReady(clientVMI, tests.LoggedInCirrosExpecter)
+			clientVMIAlternativeNamespace = tests.WaitUntilVMIReady(clientVMIAlternativeNamespace, tests.LoggedInCirrosExpecter)
 
-		// Create three vmis, vmia and vmib are in same namespace, vmic is in different namespace
-		vmia = createVMICirros(virtClient, tests.NamespaceTestDefault, map[string]string{"type": "test"})
-		vmib = createVMICirros(virtClient, tests.NamespaceTestDefault, map[string]string{})
-		vmic = createVMICirros(virtClient, tests.NamespaceTestAlternative, map[string]string{})
+			By("Start HTTP server at serverVMI on ports 80 and 81")
+			tests.StartHTTPServerWithLoggedInVMI(serverVMI, 80, false)
+			tests.StartHTTPServerWithLoggedInVMI(serverVMI, 81, false)
 
-		vmia = tests.WaitUntilVMIReady(vmia, tests.LoggedInCirrosExpecter)
-		vmib = tests.WaitUntilVMIReady(vmib, tests.LoggedInCirrosExpecter)
-		vmic = tests.WaitUntilVMIReady(vmic, tests.LoggedInCirrosExpecter)
-	})
+			By("Checking that VMIs have addresses")
+			assertVMIHasIPs(serverVMI)
```

Comment:
i think i would leave those tests as close as to the iteration on the ips,
it wont be atomic anyhow, but we have a bug atm for example that ips are dissappering,
so imo its safer to have the assert, which doesnt take much time to run, close to the iteration,
we can even create it atomic if needed by sampling the ips, making sure they are not empty and then iterating on the sampled array (it of course wont mean the scheme still have them, but at least would be more atomic and we will iterate something which isnt empty for sure)

wdyt ?

Reply:
I see theses at pre conditions for all the tests so is good to have them grouped apart from the fail fast benfict, about sampling VMI is not re-loaded after `WaitUntilVMIReady` one thing I can do though is implement `assertVMIHasIPs` with  `Eventually` + `Get` so we wait there until we have at least one ip.

Reply:
it would hide problems, we need to one shot check ips exists and to iter on those atomically

Reply:
if you create a vm, and dont do the "WaitUntilVmReady"
there wont be interfaces for example until some stage.
(it would fail here on a panic which will save us),

i think that we cant `range` something that might be nil, its in the vm scheme, which depends on kubevirt, not k8s
and we even have a real bug atm where the ips are gone (in sriov according the bug), i dont think we should take a chance that the test would actually test nothing